presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java

```diff
@@ -0,0 +1,230 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc.reader;
+
+import com.facebook.presto.memory.context.LocalMemoryContext;
+import com.facebook.presto.orc.StreamDescriptor;
+import com.facebook.presto.orc.TupleDomainFilter;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.Int128ArrayBlock;
+import com.facebook.presto.spi.type.UnscaledDecimal128Arithmetic;
+import io.airlift.slice.Slice;
+import io.airlift.slice.UnsafeSlice;
+
+import java.io.IOException;
+import java.util.Optional;
+
+import static com.facebook.presto.spi.type.UnscaledDecimal128Arithmetic.rescale;
+
+public class LongDecimalSelectiveStreamReader
+        extends AbstractDecimalSelectiveStreamReader
+{
+    public LongDecimalSelectiveStreamReader(
+            StreamDescriptor streamDescriptor,
+            Optional<TupleDomainFilter> filter,
+            boolean outputRequired,
+            LocalMemoryContext systemMemoryContext)
+    {
+        super(streamDescriptor, filter, outputRequired, systemMemoryContext);
+    }
+
+    @Override
+    int readNoFilter(int[] positions, int positionCount)
+            throws IOException
+    {
+        Slice decimal = UnscaledDecimal128Arithmetic.unscaledDecimal();
+        Slice rescaledDecimal = UnscaledDecimal128Arithmetic.unscaledDecimal();
+
+        int streamPosition = 0;
+        for (int i = 0; i < positionCount; i++) {
+            int position = positions[i];
+            if (position > streamPosition) {
+                skip(position - streamPosition);
+                streamPosition = position;
+            }
+
+            if (presentStream != null && !presentStream.nextBit()) {
+                if ((nonDeterministicFilter && filter.testNull()) || nullsAllowed) {
+                    nulls[outputPositionCount] = true;
+                    outputPositionCount++;
+                }
+            }
+            else {
+                int scale = (int) scaleStream.next();
+                dataStream.nextLongDecimal(decimal);
+                rescale(decimal, decimalType.getScale() - scale, rescaledDecimal);
+                long low = UnsafeSlice.getLongUnchecked(rescaledDecimal, 0);
+                long high = UnsafeSlice.getLongUnchecked(rescaledDecimal, Long.BYTES);
+                values[2 * outputPositionCount] = low;
+                values[2 * outputPositionCount + 1] = high;
+                if (nullsAllowed && presentStream != null) {
+                    nulls[outputPositionCount] = false;
+                }
+                outputPositionCount++;
+            }
+
+            outputPositions = positions;
```

Comment:
move this statement outside of the loop or remove it altogether

Reply:
yes its anyways init to positions